aiosmtpd/docs/proxyprotocol.rst

```diff
@@ -0,0 +1,426 @@
+
+   This is `as specified`_ in the PROXY Protocol documentation.
+
+.. _as specified: https://github.com/haproxy/haproxy/blob/1c0a722a83e7c45456a2b82c15889ab9ab5c4948/doc/proxy-protocol.txt#L176-L180
+
+.. _HAProxy2.3.0: https://github.com/haproxy/haproxy/blob/v2.3.0/doc/proxy-protocol.txt
+.. |HAProxy2.3.0| replace:: **HAProxy version 2.3.0**
+
+
+``handle_PROXY`` Hook
+=====================
+
+In addition to activating the PROXY protocol support as described above,
+you MUST implement the ``handle_PROXY`` hook.
+If the :attr:`handler` object does not implement ``handle_PROXY``,
+then all connection attempts will be rejected.
+
+The signature of ``handle_PROXY`` must be as follows:
+
+.. method:: handle_PROXY(server, session, envelope, proxy_data)
+
+   :param server: The :class:`SMTP` instance invoking the hook.
+   :type server: aiosmtpd.smtp.SMTP
+   :param session: The Session data *so far* (see Important note below)
+   :type session: Session
+   :param envelope: The Envelope data *so far* (see Important note below)
+   :type envelope: Envelope
+   :param proxy_data: The result of parsing the PROXY Header
+   :type proxy_data: ProxyData
+   :return: Truthy or Falsey, indicating if the connection may continue or not, respectively
+
+   .. important::
+
+      The ``session.peer`` attribute will contain the ``IP:port`` information
+      of the **directly adjacent** client.
+      In other word,
+      it will contain the endpoint identifier of the proxying entity.
+
+      Endpoint identifier of the "original" client will be recorded
+      *only* in the :attr:`proxy_data` parameter
+
+      The ``envelope`` data will usually be empty(ish),
+      because the PROXY handshake will take place before
+      client can send any transaction data.
+
+
+Parsing the Header
+==================
+
+You do not have to concern yourself with parsing the PROXY Protocol header;
+the ``aiosmtpd.proxy_protocol`` module contains the full parsing logic.
+
+All you need to do is to *validate* the parsed result in the ``handle_PROXY`` hook.
+
+
+``ProxyData`` API
+=================
+
+.. py:module:: aiosmtpd.proxy_protocol
+
+.. py:class:: ProxyData(\
+   version=None\
+   )
+
+   |
+   | :part:`Attributes & Properties`
+
+   .. py:attribute:: version
+      :type: Optional[int]
+
+      Contains the version of the PROXY Protocol header.
+
+      If ``None``, it indicates that parsing has failed and the header is malformed.
+
+   .. py:attribute:: command
+      :type: int
+
+      Contains the `command`_. Only set if ``version=2``
+
+   .. py:attribute:: family
+      :type: int
+
+      Contains the `address family`_. Only set if ``version=2``
+
+   .. py:attribute:: protocol
+      :type: Union[bytes, int]
+
+      For PROXY Header version 1,
+      it contains a human-readable indication of the `INET protocol and family`_.
+
+      For PROXY Header version 2,
+      it contains an integer indicating the `transport protocol being proxied`_.
```

Comment:
It's using the PROXY doc's terms [here](https://github.com/haproxy/haproxy/blob/v2.3.0/doc/proxy-protocol.txt#L207) and [here](https://github.com/haproxy/haproxy/blob/v2.3.0/doc/proxy-protocol.txt#L386).

Reply:
No, I meant the returned values so that one does not have to check the PROXY version before attempting to use the protocol:

TCP4 -> 0x11
TCP6 -> 0x21
UNKNOWN -> 0x00

Reply:
Oh okay.

That's doable.

Reply:
Done.